src/coreclr/vm/interpexec.cpp

```diff
@@ -1200,11 +1285,83 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     UNREACHABLE();
                     break;
                 }
+
+                case INTOP_ENTER_FUNCLET:
+                    // This opcode loads the exception object coming from the caller to a variable.
+                    LOCAL_VAR(ip[1], OBJECTREF) = throwable;
+                    ip += 2;
+                    break;
+
+                case INTOP_LEAVE_CATCH:
+                    // This opcode temporarily return the control back to InterpreterCodeManager::CallFunclet
+                    // The VM will resume execution on the address we return here.
+                    return (void*)(ip + 1);
+
+                case INTOP_LEAVE_FILTER:
+                    // This opcode temporarily return the control back to InterpreterCodeManager::CallFunclet
+                    // The VM will either continue the search for another exception handler or execute the handler.
+                    // depending on the result we return here.
+                    return (void*)(int64_t)((LOCAL_VAR(ip[1], int32_t) != 0) ? EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH);
+
+                case INTOP_CALL_FINALLY:
+                    {
+                        non_exceptional_finally_count += 1;
+                        LOCAL_VAR(ip[1], int8_t*) = frame;
+                        frame = stack;
+
+                        // Save current execution state for when we return from finally
+                        pFrame->ip = ip + 3;
+                        const int32_t* targetIp = ip + ip[2];
+
+                        // Allocate child frame.
+                        {
+                            InterpMethodContextFrame *pChildFrame = pFrame->pNext;
+                            if (!pChildFrame)
+                            {
+                                pChildFrame = (InterpMethodContextFrame*)alloca(sizeof(InterpMethodContextFrame));
+                                pChildFrame->pNext = NULL;
+                                pFrame->pNext = pChildFrame;
+                            }
+
+                            pChildFrame->ReInit(pFrame, pFrame->startIp, nullptr, stack + pMethod->allocaSize);
+                            pFrame = pChildFrame;
+                        }
+                        assert (((size_t)pFrame->pStack % INTERP_STACK_ALIGNMENT) == 0);
+
+                        stack = pFrame->pStack;
+                        ip = targetIp;
+                        pThreadContext->pStackPointer = stack + pMethod->allocaSize;
+                    }
+                    break;
+
+                case INTOP_RET_FINALLY:
+                    if (non_exceptional_finally_count > 0)
+                    {
+                        // This opcode returns from the finally block
+                        non_exceptional_finally_count -= 1;
```

Comment:
Will this count be properly maintained in case one of the non-exceptionally called finallys throws and the exception is caught in the same parent frame?